### PR TITLE
Translation support (i18next) + Translation DE

### PR DESCRIPTION
**Pull request by madejackson**
Hi,

I represent a distributor of a custom Raspberry Pi Solution for the professional electrician and integrator market. So far we shipped our product without a software. This can be somewhat cumbersome depending on the customer. That's why I am working on modifying cosmos-server to fit on our hardware (customized Front-End, custom Market) so we can offer it as an option (free of charge) to our customers.

As we are in a multi-language region, I need cosmos to be multilingual. Hence this PR.

I included i18next library, modified all the code to support i18next translations and added German translation already.

The translations per language all live in a single file [https://github.com/aseracorp/resiOS/tree/master/client/src/utils/locales](url), though it could be split up quiet easily into separate namespaces. I'd say ~95% of the Frontend is translated or ready for translation.

Take this PR as a base for further discussion. If you can use my contribution at all, feel free to request any changes or improvements.

### CODE REVIEW

**Comment by azukaar** [comment]
Hello, thanks for you contribution!
I will review it ASAP
Is the PR complete or do you still need time to work on it? 
thanks again!

File: client/src/utils/indexs.js [hunk 1]
```diff
@@ -1,4 +1,8 @@
 import { Button } from "@mui/material";
+import { useTranslation } from 'react-i18next';
+
+// import i18n (needs to be bundled ;))
```
Comment (mbret) [inline client/src/utils/indexs.js:4]: This comment is unclear, if something is required, it should be done in this PR. Can you explain more the purpose of this ?

File: package.json [hunk 2]
```diff
@@ -34,6 +34,9 @@
     "formik": "^2.2.9",
     "framer-motion": "^7.3.6",
     "history": "^5.3.0",
+    "i18next": "^23.11.5",
+    "i18next-fs-backend": "^2.3.1",
```
Comment (mbret) [inline package.json:38]: Maybe I missed it but it does not seem to be used ?

File: client/src/utils/i18n.jsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+const resources = {
+  en: {
+    translation: en
+  },
+  de: {
+    translation: de
+  }
+};
+
+// don't want to use this?
```
Comment (mbret) [inline client/src/utils/i18n.jsx:19]: Please remove boilerplate comments

File: client/src/utils/i18n.jsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  // detect user language
+  // learn more: https://github.com/i18next/i18next-browser-languageDetector
+  .use(LanguageDetector)
+  // pass the i18n instance to react-i18next.
+  .use(initReactI18next)
+  // init i18next
+  // for all options read: https://www.i18next.com/overview/configuration-options
+  .init({
+    resources,
+    //keySeparator: false,
```
Comment (mbret) [inline client/src/utils/i18n.jsx:37]:
```suggestion
```

File: client/src/utils/i18n.jsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    debug: true,
+    fallbackLng: 'en',
+    interpolation: {
+      escapeValue: false, // not needed for react as it escapes by default
+    }
+  });
+
+
+export default i18n;
```
Comment (mbret) [inline client/src/utils/i18n.jsx:46]: use named export, easier to maintain.

export { i18n }.

OR just export on line 23 directly.

File: client/src/utils/indexs.js [hunk 6]
```diff
@@ -52,10 +56,11 @@ export const redirectToLocal = (url) => {
 }
 
 export const crontabToText = (crontab) => {
+  const { t } = useTranslation();
```
Comment (mbret) [inline client/src/utils/indexs.js:59]: :red_circle:  Never ever use hooks in functions that are not hook. In this case it's being called in the jsx and you cannot predict the first rule of hook.

File: client/src/index.jsx [hunk 7]
```diff
@@ -6,6 +6,9 @@ import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs';
 import localizedFormat from 'dayjs/plugin/localizedFormat'; // import this for localized formatting
 import 'dayjs/locale/en-gb';
```
Comment (mbret) [inline client/src/index.jsx:7]: Can you check whether it would be possible / usefull to integrate with it ?

File: client/src/utils/i18n.jsx [hunk 8]
```diff
@@ -0,0 +1,46 @@
+import i18n from 'i18next';
```
Comment (mbret) [inline client/src/utils/i18n.jsx:1]: Move this file under /locales folder to it stay within the same feature.

**Comment by mbret** [comment]
Thanks for your work, you went the extra mile and that's appreciated. A couple of suggestions:

1. You have to namespace your keys. "New" is hard to understand and maintain. A better key is "container.created.newLabel" for exemple. We have more context, we understand it's a label and we understand it's shown in a context of a container creation. Even better, include the feature / pages or whatever else make it unique in the key. 

"management.servApps.list.refreshButton.refreshLabel"

- We understand it's the management section
- We understand it's the servApps sub section
- We understand it's the list tab (containers)
- We understand it's related to the button refresh
- We understand it's the label

Don't hesitat to be verbose, this is usually better than not.

2. Don't use deep object in the translation config. Use a flat config. It's easier to type (typescript), easier to navigate and locate. You can split your files into several if it becomes too big.

**Comment by madejackson** [comment]
Thanks @mbret @azukaar  for looking at the code. I'll try to fix the aforementioned points from @mbret in the coming days.

While working on my fork I found some issues and more text to translate so I updated the commit a couple of times.

Still on the todo is translation-support for the market. Not quiet sure how I want to do that.

I did not translate the [migrate014](https://github.com/azukaar/Cosmos-Server/blob/master/client/src/pages/home/migrate014.jsx). I think that's not nececary as this is for migrating from an english-only-version.

**Comment by mbret** [comment]
> Thanks @mbret @azukaar for looking at the code. I'll try to fix the aforementioned points from @mbret in the coming days.
> 
> While working on my fork I found some issues and more text to translate so I updated the commit a couple of times.
> 
> Still on the todo is translation-support for the market. Not quiet sure how I want to do that.
> 
> I did not translate the [migrate014](https://github.com/azukaar/Cosmos-Server/blob/master/client/src/pages/home/migrate014.jsx). I think that's not nececary as this is for migrating from an english-only-version.

I think you don't have to translate everything, as long as it suits your need at least. You already setup the fundation which is the most important point. Again thanks.

Translations can be added as we move forward and detect some missing texts.

File: client/src/utils/indexs.js (hunk 1 above)
Comment (madejackson) [reply client/src/utils/indexs.js:4]: It's copypasta from the guide: https://react.i18next.com. I'll remove it.

File: package.json (hunk 2 above)
Comment (madejackson) [reply package.json:38]: Leftover from testing. I'll remove it.

**Comment by azukaar** [comment]
For the market you mean the application? I'd say don't worry about it for now, because that needs to be done separately on the market's repo: https://github.com/azukaar/cosmos-servapps-official

**Comment by madejackson** [comment]
> Thanks for your work, you went the extra mile and that's appreciated. A couple of suggestions:
> 
>     1. You have to namespace your keys. 
> 
>     2. Don't use deep object in the translation config. Use a flat config.

@mbret Just to confirm, You prefer a flat config, mostly like it is now, but then the keys should be formatted with dots (.)?

So...
- ...I add `keySeparator: false` which disables the dot (.) as a nesting seperator to avoid object nesting
- ...this means I'll always use the full key including namespaces in the translation-files
- ...In the code I use keyPrefix while defining { t } where applicable, otherwise using the full key.

That's my plan anyhow.

**Comment by mbret** [comment]
> You prefer a flat config, mostly like it is now, but then the keys should be formatted with dots (.)?

Yes

> .this means I'll always use the full key including namespaces in the translation-files

That's the point, full intent is more clear. 

> .In the code I use keyPrefix while defining { t } where applicable, otherwise using the full key.

That is hard to maintain. Use the full key everywhere.

Remember that, the next developer should be able to understand what it is directly by looking at it. You should not have to look for a potentiall key prefix somewhere else in the code to understand which part you are working on.

The deepest the feature is in the code the longest the key will be yes, however we can probably find middleground or shortcut in the name to mitigate it. These crazy keys might still exist but nothing is perfect, you would also have a crazy deep object otherwise anyway.

Additionally and thanks to the flat structure, you can if you want introduce typing for the keys. That would be a neat addition which will prevent developers to make mistakes. Otherwise we can always have a ticket to track this for later.

File: client/src/utils/indexs.js (hunk 6 above)
Comment (madejackson) [reply client/src/utils/indexs.js:59]: I think I understand what I did wrong. I changed it to pass `t` along when calling the func.

File: client/src/index.jsx (hunk 7 above)
Comment (madejackson) [reply client/src/index.jsx:7]: Not sure what should be done here. Datetime is already formatted correctly in my language.
grafik

For timeago.js I did implement a fix, so relative time formats work as well.

**Comment by madejackson** [comment]
In the process of changing the keys. peek, give me a hint if you prefer it a bit different.

Regex used for renaming, for further reference: `((\{?t\(')|(  "))_key_(('\)\}?)|(":))`

`  "header.notification.title.certificateRenewed": "Cosmos Certificate Renewed",
  "header.notification.title.containerUpdate": "Container Update",
  "header.notification.title.alertTriggered": "Alert triggered",
  "header.notification.title.serverError": "Server Error",
  "header.notification.message.certificateRenewed": "The TLS certificate for the following domains has been renewed: {{Vars}}",
  "header.notification.message.containerUpdate": "Container {{Vars}} updated to the latest version!",
  "header.notification.message.alertTriggered": "The alert \"{{Vars}}\" was triggered.",
  "management.configuration.header.refreshButton.refreshLabel": "Refresh",
  "management.configuration.header.restartButton.restartLabel": "Restart Server",
  "management.configuration.header.purgeMetricsButton.purgeMetricsLabel": "Purge Metrics Dashboard",
  "management.configuration.header.purgeMetricsButton.purgeMetricsPopUp.cofirmAction": "Are you sure you want to purge all the metrics data from the dashboards?",
  "management.configuration.http.allowSearchIndexCheckbox": "Enable this option if you have a public site and want to allow search engines to find it, so it appears on search results. ",
  "management.configuration.http.allowSearchIndexCheckbox.allowSearchIndexLabel": "Allow search engines to index your server",
  "management.configuration.general.alert.infoConfigFile": "This page allow you to edit the configuration file. Any Environment Variable overwritting configuration won't appear here.",
  "management.configuration.general.backupDirInput.backupDirLabel": "Backup Output Directory (relative to the host server `/`)",
  "management.configuration.general.backupDirInput.backupDirHelperText": "Directory where backups will be stored (relative to the host server `/`)",
  "management.configuration.general.monitoringCheckbox.monitoringLabel": "Monitoring Enabled",
  "management.configuration.email.inbobox.label": "This allow you to setup an SMTP server for resiOS to send emails such as password reset emails and invites",
  "management.configuration.email.selfSignedCheckbox.SelfSignedHelperText": "Allow self-signed certificate",
  "management.configuration.email.selfSignedCheckbox.SelfSignedLabel": "Allow Insecure TLS",
  "management.configuration.email.tlsCheckbox.tlsLabel": "SMTP Uses TLS",
  "management.configuration.email.passwordInput.passwordLabel": "SMTP Password",
  "management.configuration.email.passwordInput.passwordHelperText": "SMTP Password",
  "management.configuration.email.usernameInput.usernameLabel": "SMTP Username",
  "management.configuration.email.usernameInput.usernameHelperText": "SMTP Username",
  "management.configuration.email.senderInput.senderLabel": "SMTP From",
  "management.configuration.email.senderInput.senderHelperText": "SMTP From",
  "management.configuration.email.enableCheckbox.enableLabel": "Enable SMTP",
  "management.configuration.email.enableCheckbox.enableHelperText": "Enable SMTP",`

**Comment by madejackson** [comment]
I think this PR is now ready to be merged. I'm gonna focus on the market now. This will be a separate PR though.

**Comment by azukaar** [comment]
Thanks for your work, I think for the market we need to talk about the solution **before** you implement anything (feel free to ping me on Discord)

**Comment by azukaar** [comment]
Hello! Had to revert the merge... I (thankfully in time) realised that you create yoru PR against MASTER instead of UNSTABLE
I typically release on the unstable beta channel before going to production
Could you re-open a PR to unstable please?

**Pull request merged**